TPP.Persistence.MongoDB/Repos/BadgeMarketRepo.cs

```diff
@@ -0,0 +1,249 @@
+                List<BadgeBuyOffer> buyOffers = await FindAllBuyOffers(null, species, null, null, shiny);
+                buyOffers = buyOffers.Where(o => o.Price >= badge.SellPrice).ToList();
+                buyOffers = buyOffers.OrderBy(o => o.WaitingSince).ToList();
+
+                if(badge.SellPrice == null)
+                    throw new InvalidOperationException("Tried to sell a badge with no sell price");
+
+                foreach (BadgeBuyOffer offer in buyOffers)
+                {
+                    if (offer.UserId == badge.UserId)
+                        continue; //user shouldn't sell to themself
+                    if (badge.UserId == null)
+                        throw new OwnedBadgeNotFoundException(badge);
+                    if (((offer.Form != null) && (offer.Form != badge.Form))
+                        || ((offer.Source != null) && (offer.Source != badge.Source)))
+                        continue;
+
+                    User? buyer = await _userRepo.FindById(offer.UserId);
+                    User? seller = await _userRepo.FindById(badge.UserId);
+                    if (buyer == null)
+                        throw new UserNotFoundException<string>(offer.UserId);
+                    if (seller == null)
+                        throw new UserNotFoundException<string>(badge.UserId);
+                    
+                    soldBadges.Add(new IBadgeMarketRepo.BadgeSale(seller, buyer, badge, (long)badge.SellPrice));
+                    await _tokenBank.PerformTransactions(
+                        new Transaction<User>[]
+                        {
+                            new Transaction<User>(buyer, -1 * offer.Price, "BadgePurchase"),
+                            new Transaction<User>(seller, offer.Price, "BadgeSale")
+                        }
+                    );
+                    await _badgeRepo.TransferBadges(new List<Badge> { badge }.ToImmutableList(), buyer.Id, "BadgeSale", new Dictionary<string, object?>() { });
+                    await ResetUserSellOffers(badge.UserId, badge.Species, badge.Form, badge.Shiny);
+                    offer.decrement(_clock.GetCurrentInstant());
+                    if (offer.Amount > 0)
+                        await BuyOfferCollection.FindOneAndReplaceAsync(o => o.Id == offer.Id, offer);
+                    else
+                        await BuyOfferCollection.FindOneAndDeleteAsync(o => o.Id == offer.Id);
+                    break; //this badge has been sold, ignore the rest of the offers
+                }
+            }
+            return soldBadges.ToImmutableList();
+        }
+        /// <summary>
+        /// Sorts badges according to the priority in which they should be sold.
+        /// Current sorting rule: prioritize keeping 1 of each form, then sell newer badges first. Top priority badge will be selected for sale when fufilling offers.
+        /// </summary>
+        private static List<Badge> SortBySpecialness(IEnumerable<Badge> toSort)
```

Comment:
I'm 95% sure with enough effort this can be expressed as a Comparer instead of an imperative sort method